File: backend/api/image_url.py
"""Normalize pasted image URLs (e.g. Google Images search/imgres links)."""
from urllib.parse import parse_qs, unquote, urlparse


def normalize_image_url(raw: str) -> str:
    """
    Turn pasted browser/Google links into a direct image URL when possible.
    Returns stripped URL or empty string.
    """
    url = (raw or "").strip()
    if not url:
        return ""

    if not url.startswith(("http://", "https://")):
        url = f"https://{url.lstrip('/')}"

    try:
        parsed = urlparse(url)
    except Exception:
        return url

    host = (parsed.netloc or "").lower()
    path = (parsed.path or "").lower()

    # Google Images result page — extract embedded image
    if "google." in host and "/imgres" in path:
        qs = parse_qs(parsed.query)
        for key in ("imgurl", "url"):
            if key in qs and qs[key]:
                return unquote(qs[key][0]).strip()

    # Google redirect wrapper
    if "google." in host and path in ("/url", "/imgres"):
        qs = parse_qs(parsed.query)
        for key in ("url", "imgurl"):
            if key in qs and qs[key]:
                candidate = unquote(qs[key][0]).strip()
                if candidate.startswith(("http://", "https://")):
                    return candidate

    return url


def image_url_help_error(url: str) -> str | None:
    """Return user-facing error if URL is unlikely to work as a hotlinked image."""
    if not url:
        return None

    parsed = urlparse(url)
    host = (parsed.netloc or "").lower()
    path = (parsed.path or "").lower()

    if "google." in host and "/search" in path:
        return (
            "That is a Google Search page link, not a direct image. "
            "Open the image in a new tab, then copy the address bar URL "
            "(should end in .jpg, .png, or contain googleusercontent.com)."
        )

    if "bing." in host and "/search" in path:
        return "Use a direct image link, not a Bing search results page."

    return None

File: frontend/src/utils/imageUrl.js
/**
 * Normalize pasted image URLs (Google imgres, etc.) to a direct image link.
 */
export function normalizeImageUrl(raw) {
  const url = (raw || "").trim();
  if (!url) return "";

  let full = url;
  if (!/^https?:\/\//i.test(full)) {
    full = `https://${full.replace(/^\/+/, "")}`;
  }

  try {
    const parsed = new URL(full);
    const host = parsed.hostname.toLowerCase();

    if (host.includes("google.") && parsed.pathname.includes("/imgres")) {
      const imgurl = parsed.searchParams.get("imgurl") || parsed.searchParams.get("url");
      if (imgurl) return decodeURIComponent(imgurl).trim();
    }

    if (host.includes("google.") && (parsed.pathname === "/url" || parsed.pathname === "/imgres")) {
      const target = parsed.searchParams.get("url") || parsed.searchParams.get("imgurl");
      if (target?.startsWith("http")) return decodeURIComponent(target).trim();
    }
  } catch {
    /* keep full */
  }

  return full;
}

export function imageUrlWarning(url) {
  if (!url) return null;
  try {
    const parsed = new URL(normalizeImageUrl(url));
    if (parsed.hostname.includes("google.") && parsed.pathname.includes("/search")) {
      return "That is a Google Search link, not a direct image. Open the image in a new tab and copy that URL.";
    }
  } catch {
    return "Invalid URL.";
  }
  return null;
}

export function imageProxyUrl(directUrl) {
  const normalized = normalizeImageUrl(directUrl);
  if (!normalized) return "";
  const base = (import.meta.env.VITE_API_BASE_URL || "/api").replace(/\/$/, "");
  return `${base}/image-proxy/?url=${encodeURIComponent(normalized)}`;
}
